clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3430,147 +3498,134 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
+    auto *SIMWGSA = cast<SYCLIntelMaxWorkGroupSizeAttr>(A);
+    if (auto *Existing = SYCLKernel->getAttr<ReqdWorkGroupSizeAttr>()) {
+      ASTContext &Ctx = S.getASTContext();
+      if (Existing->getXDimVal(Ctx) > SIMWGSA->getXDimVal(Ctx) ||
+          Existing->getYDimVal(Ctx) > SIMWGSA->getYDimVal(Ctx) ||
+          Existing->getZDimVal(Ctx) > SIMWGSA->getZDimVal(Ctx)) {
+        S.Diag(SYCLKernel->getLocation(),
+               diag::err_conflicting_sycl_kernel_attributes);
+        S.Diag(Existing->getLocation(), diag::note_conflicting_attribute);
+        S.Diag(SIMWGSA->getLocation(), diag::note_conflicting_attribute);
+        SYCLKernel->setInvalidDecl();
+      } else {
+        SYCLKernel->addAttr(A);
       }
+    } else {
+      SYCLKernel->addAttr(A);
     }
+    break;
+  }
+  case attr::Kind::SYCLIntelKernelArgsRestrict:
+  case attr::Kind::SYCLIntelNumSimdWorkItems:
+  case attr::Kind::SYCLIntelSchedulerTargetFmaxMhz:
+  case attr::Kind::SYCLIntelMaxGlobalWorkDim:
+  case attr::Kind::SYCLIntelNoGlobalWorkOffset:
+  case attr::Kind::SYCLIntelUseStallEnableClusters:
+  case attr::Kind::SYCLIntelLoopFuse:
+  case attr::Kind::SYCLIntelFPGAMaxConcurrency:
+  case attr::Kind::SYCLIntelFPGADisableLoopPipelining:
+  case attr::Kind::SYCLIntelFPGAInitiationInterval:
+  case attr::Kind::SYCLSimd: {
+    if ((A->getKind() == attr::Kind::SYCLSimd) && KernelBody &&
+        !KernelBody->getAttr<SYCLSimdAttr>()) {
+      // Usual kernel can't call ESIMD functions.
+      S.Diag(KernelBody->getLocation(),
+             diag::err_sycl_function_attribute_mismatch)
+          << A;
+      S.Diag(A->getLocation(), diag::note_attribute);
+      KernelBody->setInvalidDecl();
+    } else
+      SYCLKernel->addAttr(A);
+    break;
+  }
+  // TODO: vec_len_hint should be handled here
+  default:
+    // Seeing this means that CollectPossibleKernelAttributes was
+    // updated while this switch wasn't...or something went wrong
+    llvm_unreachable("Unexpected attribute was collected by "
+                     "CollectPossibleKernelAttributes");
   }
-  for (const auto &elt : Marker.KernelSet) {
-    if (FunctionDecl *Def = elt->getDefinition())
-      Marker.TraverseStmt(Def->getBody());
+}
+
+void Sema::MarkDevices(void) {
+  // This Tracker object ensures that the SyclDeviceDecls collection includes
+  // the SYCL_EXTERNAL functions, and manages the diagnostics for all of the
+  // functions in the kernel.
+  DeviceFunctionTracker Tracker(*this);
+
+  for (Decl *D : syclDeviceDecls()) {
+    // TODO: ERICH: original code had a dyn_cast here, do we know if a
+    // non-function can end up in the sycl kernel list?  I don't think so...
+    FunctionDecl *SYCLKernel = cast<FunctionDecl>(D);
+
+    // This type does the actual analysis on a per-kernel basis. It does this to
+    // make sure that we're only ever dealing with the context of a single
+    // kernel at a time.
+    SingleDeviceFunctionTracker T{Tracker, SYCLKernel};
```

Comment:
I found we were constantly doing a set/reset of the collections owned by the parent when we were at a point where we only cared about the current kernel.  This object does the analysis work, and just rejoins its information into the parent collection when it is done.